verl/workers/rollout/schemas.py

```diff
@@ -82,127 +88,121 @@ class AsyncRolloutRequest(BaseModel):
+            raise ValueError("max_prompt_len is required for AsyncRolloutRequest initialization")
+        if not (tokenizer := values.pop("tokenizer", None)):
+            raise ValueError("tokenizer is required for AsyncRolloutRequest initialization")
+
+        values["messages"] = [Message.model_validate(msg) for msg in messages]
+        values["messages_dumps"] = [msg.model_dump() for msg in values["messages"]]
+
+        if tool_schemas := values.get("tool_schemas"):
+            tools = values["tools"] = [tool.model_dump() for tool in tool_schemas]
+            tokens_without_prompt = tokenizer.apply_chat_template(messages, tools=tools, add_generation_prompt=False, tokenize=True)
+            tokenization_dict_with_prompt = tokenizer.apply_chat_template(messages, tools=tools, add_generation_prompt=True, tokenize=True, return_dict=True)
+            values["input_ids"] = values["prompt_ids"] = tokenization_dict_with_prompt["input_ids"]
+            values["attention_mask"] = values["prompt_attention_mask"] = tokenization_dict_with_prompt["attention_mask"]
+            if len(values["input_ids"]) > max_prompt_len:
+                # Only log the warning to avoid truncating in the middle of generation prompt. Consider raising an error for this case in the future.
+                logger.warning(f"Prompt {values['batch_data_id']} length {len(values['input_ids'])} greater than max_prompt_len {max_prompt_len} after applied chat template with tools.")
+        elif not values.get("input_ids") or not values.get("attention_mask"):
+            raise ValueError("input_ids and attention_mask is required for requests without tools")
+        else:
+            tokens_without_prompt = tokenizer.apply_chat_template(messages, add_generation_prompt=False, tokenize=True)
+            values["prompt_ids"], values["prompt_attention_mask"] = values["input_ids"], values["attention_mask"]
+
+        values["position_ids"] = values["prompt_position_ids"] = compute_position_id_with_mask(torch.tensor(values["attention_mask"])).tolist()
+        values["loss_mask"] = values["prompt_loss_mask"] = [0] * len(values["input_ids"])
+        values["generation_prompt_ids"] = values["input_ids"][len(tokens_without_prompt) :]
+        return values
+
+    def _update_input_ids(self, new_input_ids: List[int], attention_mask: bool, loss_mask: bool, full_tokens: bool = False) -> None:
+        message_len_delta = (len(new_input_ids) - len(self.input_ids)) if full_tokens else len(new_input_ids)
+        self.input_ids = new_input_ids if full_tokens else (self.input_ids + new_input_ids)
+        attention_mask = [int(attention_mask)] * message_len_delta
+        self.attention_mask += attention_mask
+        _delta_position_ids = compute_position_id_with_mask(torch.tensor(attention_mask)).tolist()
+        last_position_id = self.position_ids[-1]
+        _position_ids = [pos_id + last_position_id for pos_id in _delta_position_ids]
```

Comment:
_position_ids = (_delta_position_ids + (last_position_id + 1)).tolist() may be a good fix. Before impl I made a bug here, but won't trigger due to regenerated position_ids when build batch. https://github.com/volcengine/verl/blob/main/verl/workers/rollout/sglang_rollout/async_sglang_rollout.py#L413